File: 2131-longest-palindrome-by-concatenating-two-letter-words/2131-longest-palindrome-by-concatenating-two-letter-words.py
class Solution:
    def longestPalindrome(self, words: List[str]) -> int:
        count = Counter(words)
        length = 0
        has_center = False # 用來記錄是否有可以放在中間的字串
        
        for word in count:
            rev = word[::-1]
            if word != rev:
                if rev in count:
                    pair = min(count[word], count[rev])
                    length += pair * 4
                    # 將配對用掉，避免重複使用
                    count[word] -= pair
                    count[rev] -= pair
            else:
                # 自己就是回文（如 "gg"）
                pairs = count[word] // 2
                length += pairs * 4
                count[word] -= pairs * 2
                if count[word] > 0:
                    has_center = True
        
        if has_center:
            length += 2  # 加一組可以放在中間的回文

        return length

File: 2131-longest-palindrome-by-concatenating-two-letter-words/2131-longest-palindrome-by-concatenating-two-letter-words.cpp
class Solution {
public:
    int longestPalindrome(vector<string>& words) {
        unordered_map<string, int> count;
        int length = 0;
        bool has_center = false;

        // 統計每個字串出現次數
        for (const string& word : words) {
            count[word]++;
        }

        for (auto& [word, freq] : count) {
            string rev = word;
            reverse(rev.begin(), rev.end());

            if (word != rev) {
                // 如果有對應的反轉字，進行配對
                if (count.find(rev) != count.end()) {
                    int pair = min(freq, count[rev]);
                    length += pair * 4;
                    count[word] -= pair;
                    count[rev] -= pair;
                }
            } else {
                // 自己就是回文（例如 "gg"）
                int pair = freq / 2;
                length += pair * 4;
                count[word] -= pair * 2;

                if (count[word] > 0) {
                    has_center = true;
                }
            }
        }

        if (has_center) {
            length += 2; // 可放在中心的回文
        }

        return length;
    }
};